internal/juno/juno.go

```diff
@@ -0,0 +1,254 @@
+		n.metricsServer.ListenAndServe(metricsErrCh)
+	}
+
+	if err = <-rpcErrCh; err != nil {
+		return err
+	}
+
+	if n.metricsServer != nil {
+		if err = <-metricsErrCh; err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (n *Node) Shutdown() error {
+	n.stateManager.Close()
+	n.transactionManager.Close()
+	n.blockManager.Close()
+	n.stateManager.Close()
+	n.synchronizer.Close()
+
+	if err := n.rpcServer.Close(shutdownTimeout); err != nil {
+		return err
+	}
+
+	if n.metricsServer != nil {
+		if err := n.metricsServer.Close(shutdownTimeout); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func starkNetJsonRPC(stateManager *state.Manager, blockManager *block.Manager,
+	transactionManager *transaction.Manager, synchronizer *syncer.Synchronizer,
+	virtualMachine *cairovm.VirtualMachine,
+) (*jsonrpc.JsonRpc, error) {
+	starkNetApi := starknet.New(stateManager, blockManager, transactionManager, synchronizer,
+		virtualMachine)
+	jsonRpc := jsonrpc.NewJsonRpc()
+	handlers := []struct {
+		name       string
+		function   any
+		paramNames []string
+	}{
+		{"starknet_getBlockWithTxHashes", starkNetApi.GetBlockWithTxHashes, []string{"block_id"}},
+		{"starknet_getBlockWithTxs", starkNetApi.GetBlockWithTxs, []string{"block_id"}},
+		{"starknet_getStateUpdate", starkNetApi.GetStateUpdate, []string{"block_id"}},
+		{"starknet_getStorageAt", starkNetApi.GetStorageAt, []string{"block_id", "contract_address", "key"}},
+		{"starknet_getTransactionByHash", starkNetApi.GetTransactionByHash, []string{"transaction_hash"}},
+		{"starknet_getTransactionByBlockIdAndIndex", starkNetApi.GetTransactionByBlockIdAndIndex, []string{"block_id", "index"}},
+		{"starknet_getTransactionReceipt", starkNetApi.GetTransactionReceipt, []string{"transaction_hash"}},
+		{"starknet_getClass", starkNetApi.GetClass, []string{"class_hash"}},
+		{"starknet_getClassHashAt", starkNetApi.GetClassHashAt, []string{"block_id", "address"}},
+		{"starknet_getBlockTransactionCount", starkNetApi.GetBlockTransactionCount, []string{"block_id"}},
+		{"starknet_call", starkNetApi.Call, []string{"block_id", "request"}},
+		{"starknet_estimateFee", starkNetApi.EstimateFee, []string{"block_id", "request"}},
+		{"starknet_blockNumber", starkNetApi.BlockNumber, nil},
+		{"starknet_blockHashAndNumber", starkNetApi.BlockHashAndNumber, nil},
+		{"starknet_chainId", starkNetApi.ChainId, nil},
+		{"starknet_pendingTransactions", starkNetApi.PendingTransactions, nil},
+		{"starknet_protocolVersion", starkNetApi.ProtocolVersion, nil},
+		{"starknet_syncing", starkNetApi.Syncing, nil},
+	}
+	for _, handler := range handlers {
+		if err := jsonRpc.RegisterFunc(handler.name, handler.function, handler.paramNames...); err != nil {
+			return nil, err
+		}
+	}
+	return jsonRpc, nil
+}
+
+func defaultDbPathPrefix() (string, error) {
+	junoDir := ".juno"
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(homeDir, junoDir), nil
+}
```

Comment:
The reason I didn't like `UserDataDir` is that it is doing 2 things at the same time. The function should not be creating the directory and there is a bug in the function: it should be using `os.MkdirAll` instead of `os.Mkdir`.

I was following go-ethereum's convention of storing the data information under `$HOME/.juno`

Reply:
Also using `UserDataDir` will break the tests.

Reply:
```go
func defaultDbPathPrefix() (string, error) {
	const junoDir = ".juno"

	switch runtime.GOOS {
	case "windows":
		// On Windows ConfigDir and DataDir share the same path. See:
		// https://stackoverflow.com/questions/43853548/xdg-basedir-directories-for-windows.
		configDir, err := os.UserConfigDir()
		if err != nil {
			return "", err
		}
		return filepath.Join(configDir, junoDir), nil
	case "darwin", "dragonfly", "freebsd", "illumos", "ios", "linux", "netbsd",
		"openbsd", "solaris":
		// Use XDG_DATA_HOME. If it is not configured, try $HOME/.local/share
		// See https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
		dataHome := os.Getenv("XDG_DATA_HOME")
		if dataHome != "" {
			return filepath.Join(dataHome, junoDir), nil
		}
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, junoDir), nil
}
```

The above follows the requirements, however, it is slightly more difficult to test. Let me know what you think

Reply:
Maybe the path is split the two methods, one that gives you the config path and the other that creates the file.

Agree with @joshklop that we should follow standard XDG, not just because Geth does it in that way

Reply:
> The reason I didn't like `UserDataDir` is that it is doing 2 things at the same time. The function should not be creating the directory 

As @stdevMac said, we can split the function in two or rename it.

> There is a bug in the function: it should be using `os.MkdirAll` instead of `os.Mkdir`.

That is by design. By my interpretation of the spec, the user should have the `~/.local/share` directory already. Juno is not responsible for creating it. I don't have serious problems with changing this, though.

> The above follows the requirements, however, it is slightly more difficult to test. Let me know what you think

I think it looks good, with a few caveats for spec compliance:

* We need to [fall back](https://github.com/NethermindEth/juno/blob/7996505200e0087d33dc15959c733a1b77b5830c/internal/config/config.go#L77-L88) to `~/.local/share` in case `XDG_DATA_DIR` is not defined.
* The `junoDir` should be `juno`, not `.juno`.

Reply:
```go
func defaultDbPathPrefix() (string, error) {
	const local, share, junoDir = ".local", "share", "juno"

	switch runtime.GOOS {
	case "windows":
		// On Windows ConfigDir and DataDir share the same path. See:
		// https://stackoverflow.com/questions/43853548/xdg-basedir-directories-for-windows.
		configDir, err := os.UserConfigDir()
		if err != nil {
			return "", err
		}
		return filepath.Join(configDir, junoDir), nil
	case "darwin", "dragonfly", "freebsd", "illumos", "ios", "linux", "netbsd",
		"openbsd", "solaris":
		// See https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
		dataHome := os.Getenv("XDG_DATA_HOME")
		if dataHome != "" {
			return filepath.Join(dataHome, junoDir), nil
		}
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, local, share, junoDir), nil
}
```

Does the above look good?

Reply:
I think so. Thanks!

Reply:
I have had to split the above function into 2 separate functions for testing purposes. Please have a look at the latest commit and please resolve the conversation.

Reply:
LGTM

Reply:
What do you think @joshklop ?